plugin-base/src/main/java/io/everytrade/server/plugin/impl/everytrade/KuCoinDownloader.java

```diff
@@ -8,41 +8,489 @@
+            && state.oldLastTradeEndDate == EXCHANGE_START_DATE.getTime()) {
+            long newStartDate;
+            long newEndDate;
+            // case very first download
+            if (state.newLastTradeEndDate == 0 && state.newLastTradeStartDate == 0) {
+                newEndDate = now;
+                newStartDate = newEndDate - WEEK;
+                if (newStartDate < state.oldFirstTradeDate) {
+                    newStartDate = state.oldFirstTradeDate;
+                }
+            }
+            // case first download has finished and start new one
+            else if (state.newLastTradeStartDate == state.newLastTradeEndDate) {
+                state.oldFirstTradeDate = state.newLastTradeStartDate;
+                newEndDate = now;
+                newStartDate = newEndDate - WEEK;
+                if (newStartDate < state.oldFirstTradeDate) {
+                    newStartDate = state.oldFirstTradeDate;
+                }
+            // case not finished download
+            } else if (state.newLastTradeStartDate > 0 && state.newLastTradeEndDate > 0
+                && state.newLastTradeStartDate != state.newLastTradeEndDate) {
+                newEndDate = state.newLastTradeStartDate;
+                newStartDate = newEndDate - WEEK;
+                if (newStartDate < state.oldFirstTradeDate) {
+                    newStartDate = state.oldFirstTradeDate;
+                }
+            } else {
+                throw new IllegalStateException("User trade history download failed. ");
+            }
+
+            while (sentRequests < MAX_TRADE_REQUEST_COUNT) {
+                ++sentRequests;
+                // This API is restricted for each account, the request rate limit is 9 times/3s.
+                Thread.sleep(SLEEP_BETWEEN_REQUESTS);
+                final List<UserTrade> userTradesBlock;
+                try {
+                    params.setEndTime(new Date(newEndDate));
+                    params.setStartTime(new Date(newStartDate));
+                    params.setNextPageCursor(String.valueOf(PAGE));
+                    userTradesBlock = tradeService.getTradeHistory(params).getUserTrades();
+                } catch (Exception e) {
+                    throw new IllegalStateException("User trade history download failed. ", e);
+                }
+                // case too much txs
+                if (userTradesBlock.size() == PAGE_LIMIT) {
+                    state.newLastTradeEndDate = newEndDate;
+                    state.newLastTradeStartDate = newStartDate;
+                    newEndDate = userTradesBlock.get(0).getTimestamp().getTime();
+                    userTrades.addAll(userTradesBlock);
+                    continue;
+                }
+                // case too much txs
+                if (userTradesBlock.size() > PAGE_LIMIT) {
+                    throw new IllegalStateException(String.format("User trade history download failed.  Txs in response exceeds limit %s",
```

Comment:
this is not nicely formated. You should put all params to new line or none. What about this : 

```
throw new IllegalStateException(
    String.format("User trade history download failed.  Txs in response exceeds limit %s", PAGE_LIMIT)
);
```